front/src/applications/stdcm/utils.ts

```diff
@@ -36,5 +40,60 @@ export function formatCreationDate(date: string) {
 
 export function extractSpeedLimit(speedLimitByTag: string): string {
   const parts = speedLimitByTag.split(' - ');
-  return parts.length > 1 ? parts[1] : speedLimitByTag;
+  return parts[parts.length - 1];
+}
+
+export function getOperationalPointsWithTimes(simulationReport: SimulationReportSheetProps): {
+  opId: string;
+  positionOnPath: number;
+  time: string | null;
+  name: string | undefined;
+  ch: string | undefined;
+  stop: string | null | undefined;
+  departureTime: string;
+}[] {
+  const operationalPoints = simulationReport.pathProperties?.suggestedOperationalPoints || [];
+  const { simulation } = simulationReport.stdcmData;
+
+  if (simulation.status !== 'success') {
+    throw new Error('Simulation was not successful.');
+  }
+
+  const { positions, times } = simulation.final_output;
+  const departureTime = extractTime(simulationReport.stdcmData.departure_time);
+
+  // Parse departure time into hours and minutes
+  const [departureHour, departureMinute] = departureTime.split(':').map(Number);
+
+  // Function to add minutes to the departure time
+  const addMinutesToTime = (baseHour: number, baseMinute: number, minutesToAdd: number) => {
+    const totalMinutes = baseMinute + minutesToAdd;
+    const hoursToAdd = Math.floor(totalMinutes / 60);
+    const finalMinutes = totalMinutes % 60;
+    const finalHour = (baseHour + hoursToAdd) % 24;
+    return `${String(finalHour).padStart(2, '0')}:${String(finalMinutes).padStart(2, '0')}`;
+  };
+
+  // Map operational points with their positions and times
+  const opResults = operationalPoints.map((op) => {
+    const index = positions.findIndex((position) => position >= op.positionOnPath);
+    const timeInMillis = index !== -1 ? times[index] : null;
+    const timeInMinutes = timeInMillis !== null ? Math.floor(timeInMillis / 60000) : null;
+    const formattedTime =
+      timeInMinutes !== null
+        ? addMinutesToTime(departureHour, departureMinute, timeInMinutes)
+        : null;
+
+    return {
+      opId: op.opId,
+      positionOnPath: op.positionOnPath,
+      time: formattedTime,
+      name: op.name,
+      ch: op.ch,
+      stop: op.stopFor,
+      departureTime,
+    };
+  });
```

Comment:
It may be mor readdable and testable if you split abit your code like this
```
const addMinutesToTime = (baseHour: number, baseMinute: number, minutesToAdd: number): string => {
  const totalMinutes = baseHour * 60 + baseMinute + minutesToAdd;
  const finalHour = Math.floor(totalMinutes / 60) % 24;
  const finalMinutes = totalMinutes % 60;
  return `${String(finalHour).padStart(2, '0')}:${String(finalMinutes).padStart(2, '0')}`;
};

const getTimeAtPosition = (
  position: number,
  positions: number[],
  times: number[],
  departureHour: number,
  departureMinute: number
): string | null => {
  const index = positions.findIndex(pos => pos >= position);
  if (index === -1) return null;

  const timeInMillis = times[index];
  const timeInMinutes = Math.floor(timeInMillis / 60000);
  return addMinutesToTime(departureHour, departureMinute, timeInMinutes);
};

// Map operational points with their positions and times
const opResults = operationalPoints.map(op => {
  const formattedTime = getTimeAtPosition(op.positionOnPath, positions, times, departureHour, departureMinute);

  return {
    opId: op.opId,
    positionOnPath: op.positionOnPath,
    time: formattedTime,
    name: op.name,
    ch: op.ch,
    stop: op.stopFor,
    departureTime,
  };
});
```

Reply:
Nice catch, done